src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,399 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Internal.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    public static class ReadOnlyMemoryUtils
+    {
+
+        /// <summary>
+        /// This implements IEquatable's Equals method.
+        /// </summary>
+        public static bool Equals(ReadOnlyMemory<char> b, ReadOnlyMemory<char> memory)
+        {
+            if (memory.Length != b.Length)
+                return false;
+
+            Contracts.Assert(memory.IsEmpty == b.IsEmpty);
+
+            for (int i = 0; i < memory.Length; i++)
+            {
+                if (memory.Span[i] != b.Span[i])
```

Comment:
>Span [](start = 27, length = 4)

Please store the spans as a variable upfront, then operate on them. These calls to `.Span` property are not free. Here and everywhere.

Consider this code.

```csharp
class Program
{
    private static void TimeTest(string str)
    {
        long temp = 0;

        var sw = new Stopwatch();
        const int trials = 1_000_000_000;

        sw.Restart();
        for (int i = 0; i < trials; ++i)
            temp += str[0];
        sw.Stop();
        Console.WriteLine($"{sw.Elapsed} on string");
        temp = 0;

        sw.Restart();
        var mem = str.AsMemory();
        for (int i = 0; i < trials; ++i)
            temp += mem.Span[0];
        sw.Stop();
        Console.WriteLine($"{sw.Elapsed} on memory.Span");
        temp = 0;

        sw.Restart();
        var span = mem.Span;
        for (int i = 0; i < trials; ++i)
            temp += span[0];
        sw.Stop();
        Console.WriteLine($"{sw.Elapsed} on a stored span");
    }

    private static void Main(string[] args)
    {
        TimeTest("hello");
    }
}
```

When compiled and run in .NET Core 2.1, in a release build, the timings are this:

```
00:00:00.4667924 on string
00:00:01.8801733 on memory.Span
00:00:00.2327934 on a stored span
```